### PR TITLE
wip: refactor first draft

### PR DESCRIPTION
**Pull request by zeeshanlakhani**

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -12,34 +12,34 @@ repos:
...
+        types: [rust]
+        pass_filenames: false
+      - id: clippy
+        name: clippy
+        description: Lint via clippy
+        entry: cargo clippy
+        language: system
+        args: ["--", "-D", "warnings"]
+        types: [rust]
+        pass_filenames: false
```
Comment (expede) [inline .pre-commit-config.yaml:29]: haha yeah I guess I can just run `--no-verify`. It is nice to have the rest of them run while in WIP, though.

File: src/main.rs [hunk 2]
```diff
@@ -1,245 +1,202 @@
...
+    io::{self, Cursor, Write},
+    str::{self, FromStr},
 };
 use url::Url;
 use wasmer::{imports, Function, Instance, Module, Store, Type, Value};
 
-/// Main entry point.
 #[tokio::main]
-async fn main() -> Result<(), Box<dyn Error>> {
+async fn main() -> Result<()> {
```
Comment (expede) [inline src/main.rs:34]: Oh yeah, that's way easier 😅

File: src/main.rs [hunk 3]
```diff
@@ -1,245 +1,202 @@
...
-    let file_content = match file_map {
-        Ok(x) => x.0,
-        Err(y) => panic!("no file {}", y),
-    };
+        Argument::Get { name } => {
+            let providers = client.get_providers(name.clone()).await?;
+            let providers = providers
+                .is_empty()
+                .then_some(providers)
+                .ok_or_else(|| anyhow!("Could not find provider for file {name}."))?;
```
Comment (expede) [inline src/main.rs:70]: Why the reassignment? Can this not be...

```rust
let providers =
  client
    .get_providers(name.clone())
    .await?
    .is_empty()
    .then_some(providers)
    .ok_or_else(|| anyhow!("Could not find provider for file {name}."))?;
```

...or is that unidiomatic?

File: src/main.rs [hunk 4]
```diff
@@ -1,245 +1,202 @@
...
-            })
-            .collect(),
-    );
-
-    let closure = workflow::closure::Closure {
-        resource,
-        action: workflow::closure::Action::from("wasm/run"),
-        inputs: workflow::closure::Input::IpldData(ipld_args),
-    };
+            println!("Fuck: {providers:?}");
```
Comment (expede) [inline src/main.rs:72]: Just a `fixme`?

File: src/main.rs [hunk 5]
```diff
@@ -1,245 +1,202 @@
...
-        match network_events.next().await {
-            Some(network::Event::InboundRequest { request, channel }) => {
-                if request == closure_cid2 {
-                    network_client.respond_file(res_copy.clone(), channel).await;
-                }
-            }
-            e => todo!("{:?}", e),
+            //
+        }
+        Argument::Provide { wasm, fun, args } => {
```
Comment (expede) [inline src/main.rs:89]: This feels like a big block to just put inside a branch, no? Is there a cost to breaking this up and/or why not move it to another module?

File: src/main.rs [hunk 6]
```diff
@@ -1,245 +1,202 @@
...
+                    .map(|a| match a {
+                        Value::I32(i) => Ipld::from(*i),
+                        _ => todo!(),
+                    })
+                    .collect(),
+            );
+
+            let closure = Closure {
+                resource,
+                action: Action::try_from("wasm/run")?,
```
Comment (expede) [inline src/main.rs:156]: Why `try_from`? AFAICT the smart constructor should be unfailable, no?

File: src/workflow/receipt.rs [hunk 7]
```diff
@@ -1,5 +1,5 @@
 use crate::db::schema::receipts;
-use diesel::prelude::*;
+use diesel::{Insertable, Queryable};
```
Comment (expede) [inline src/workflow/receipt.rs:2]: 👍

File: src/workflow/task.rs [hunk 8]
```diff
@@ -1,34 +1,42 @@
-//! A [Closure] wrapped with [Config]uration, metadata, and so on.
+//! A [Closure] wrapped with Configuration, metadata, and optional secret.
+
 use crate::workflow::{closure::Closure, config::Resources};
-use libipld::Ipld;
+use libipld::{serde as ipld_serde, Ipld};
```
Comment (expede) [inline src/workflow/task.rs:4]: Just shorter than `libipld::serde`?

File: src/workflow/task.rs [hunk 9]
```diff
@@ -1,34 +1,42 @@
...
 use std::collections::BTreeMap;
 
+const RESOURCES_KEY: &str = "resources";
+const META_KEY: &str = "meta";
+const SECRETS_KEY: &str = "secret";
+
+const WITH_KEY: &str = "with";
+const DO_KEY: &str = "do";
+const INPUTS_KEY: &str = "inputs";
```
Comment (expede) [inline src/workflow/task.rs:13]: 👍 yeah good idea

Comment (expede) [reply src/workflow/task.rs:13]: I feel like I'm converting between string types a lot 🤔 Is defaulting to `String` most places the wrong option?

File: src/workflow/task.rs [hunk 10]
```diff
@@ -37,83 +45,46 @@ impl From<Closure> for Task {
...
-                    assoc.get("secret").ok_or(()).and_then(|ipld| match ipld {
-                        Ipld::Bool(b) => Ok(Some(*b)),
-                        Ipld::Null => Ok(None),
-                        _ => Err(()),
-                    })?;
+        let resources = if let Some(ipld) = map.get(RESOURCES_KEY) {
+            Resources::try_from(ipld)?
+        } else {
+            Resources::default()
+        };
```
Comment (expede) [inline src/workflow/task.rs:79]: `let resources = if let Some(ipld) =`

Woah, that's gnarly... but idiomatic?

Could it be rephrased as something like...

```rust
let resources = map
    .get(RESOURCES_KEY)
    .map(Resources::try_from)
    .unwrap_or_else(Resources::default);
```

...or is that not commonly done?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (zeeshanlakhani) [reply .pre-commit-config.yaml:29]: yep, but we'll get these working normally too (soon)

File: src/main.rs (hunk 3 above)
Comment (zeeshanlakhani) [reply src/main.rs:70]: there was a reason during cleanup, but I think you're right that it flows through :). I'll do this now.

File: src/main.rs (hunk 4 above)
Comment (zeeshanlakhani) [reply src/main.rs:72]: Z'isms,  yes!

File: src/main.rs (hunk 5 above)
Comment (zeeshanlakhani) [reply src/main.rs:89]: So, I'll put a todo here. I don't want the big block, but I was rethinking ways to abstract this out. So, I put it back in one spot before coming back to it.

File: src/main.rs (hunk 6 above)
Comment (zeeshanlakhani) [reply src/main.rs:156]: b/c this is fromstr under the hood, and the canonical approach seems to be tryfrom for this (and use fromstr at the bottom)

File: src/workflow/closure.rs [hunk 11]
```diff
@@ -61,26 +60,28 @@ impl Into<Ipld> for Closure {
 }
 
 impl TryFrom<Ipld> for Closure {
-    type Error = ();
+    type Error = anyhow::Error;
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply src/workflow/closure.rs:63]: I think it's good to leave this here b/c it cleans up results and error handling immensely elsewhere. We can then move to more specific errors later (via this error, or conversions to anyhow). Ideally, anyhow is the standard.

File: src/workflow/task.rs (hunk 8 above)
Comment (zeeshanlakhani) [reply src/workflow/task.rs:4]: So, pattern from wnfs :).

File: src/workflow/task.rs (hunk 9 above)
Comment (zeeshanlakhani) [reply src/workflow/task.rs:13]: it's a Rust thing. Going w/ String, means no consts and literals anyway sadly.

File: src/workflow/task.rs [hunk 12]
```diff
@@ -37,83 +45,46 @@ impl From<Closure> for Task {
         Task {
             closure,
             metadata: Ipld::Null,
-            resources: Resources {
-                fuel: None,
-                time: None,
-            },
+            resources: Resources::default(),
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply src/workflow/task.rs:48]: def not a codesmell in Rust land and is common across most structs.

[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply src/workflow/task.rs:48]: passing around None params is typically more "frowned upon".

File: src/workflow/task.rs (hunk 10 above)
Comment (zeeshanlakhani) [reply src/workflow/task.rs:79]: @expede yep, this is better (had this in the about to push)

File: src/workflow/closure.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (expede) [reply src/workflow/closure.rs:63]: Is that so we're not converting between error types and wrapping them in aggregate errors and whatnot?

File: src/workflow/task.rs (hunk 10 above)
Comment (zeeshanlakhani) [reply src/workflow/task.rs:79]: even cleaner:

```rust
let resources = map
            .get(RESOURCES_KEY)
            .map_or_else(|| Ok(Resources::default()), Resources::try_from)?;
```

File: src/main.rs (hunk 6 above)
Comment (expede) [reply src/main.rs:156]: Huh, okay. Seems wasteful, but sure 🤷‍♀️

Comment (zeeshanlakhani) [reply src/main.rs:156]: actually, yeah, the normal From is fine (i was pre-optimizing). will return.

File: src/main.rs (hunk 3 above)
Comment (expede) [reply src/main.rs:70]: I certainly have a lot of dangling stuff like this in my current code as well! Not a criticism or anything; just letting me inner 5 year old out. "Why? Why? Why? Why?" 😁

File: .pre-commit-config.yaml (hunk 1 above)
Comment (expede) [reply .pre-commit-config.yaml:29]: Just an FYI, my branches will probably have a bunch of `--no-verify` while they're WIP. I tend to push before e.g. a string of meetings, and a lot of those states won't compile.

I think Quinn has also disabled a bunch of these checks for the same reason.

Comment (expede) [reply .pre-commit-config.yaml:29]: Would it make sense to push some of these to a GH Action?

Comment (zeeshanlakhani) [reply .pre-commit-config.yaml:29]: I look at clippy as another extension of the type system :).

File: src/main.rs (hunk 3 above)
Comment (zeeshanlakhani) [reply src/main.rs:71]: @expede oh yeah, this won't work. That's due to how then_some expects the input.

File: src/workflow/closure.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply src/workflow/closure.rs:63]: yeah. you essentially want to use anyhow as your ground error type.

**Review by expede — APPROVED** [review]
Looks good to me! 🚀 Thanks!

**Pull request merged**